text/0000-private-registry-auth.md

```diff
@@ -0,0 +1,130 @@
+modifying the registry definition in the index:
+```json
+{
+    "dl": "https://my-intranet.local/api/v1/crates",
+    "api": "https://my-intranet.local/",
+    "auth_required": true,
+    "token_url": "https://my-intranet.local/me"
+}
+```
+
+This will cause Cargo to send the token obtained via `cargo login` to the
+registry with all requests. If not specified, will default to `false`.
+`cargo login` will prompt the user to obtain a token via the URL in `token_url`
+if specified, otherwise the `api` URL plus `/me`.
+
+It is up to the registry to decide whether authentication should be required for
+the Git index. If authentication is required, the registry should provide
+instructions on how to authenticate with Git along with how to add the registry
+to Cargo.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The variable `auth_required` is a boolean that enables the sending of the token
+in the `Authorization` header on all requests, including downloads and API
+calls. This will be implemented with a check on any requests to the registry
+that sends the header if this value is true. This value defaults to `false`.
+
+The `token_url` variable is used to tell Cargo where to direct the user to get
+their token. If not specified, this will default to the value in `api` plus
+`/me`.
+
+<!-- This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work. -->
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This is probably biased, but I know of no drawbacks.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This design may not be the best design, so feedback is appreciated.
+
+Another considered design was to use a cargo plugin to proxy requests through a
+local server that sends authentication with the server with requests. This
+design was decided against due to concerns about battery life, and the
+requirement to ensure the proxy is running whenever it is required, meaning
+possibly requiring a system service to keep it running.
+
+The impact of not doing this would be that large organizations would be hesitant
+to use Rust for internal/confidential projects, so there would not be as many
+large organizations supporting the growth of Rust.
+
+<!-- - Why is this design the best in the space of possible designs?
+- What other designs have been considered and what is the rationale for not choosing them?
+- What is the impact of not doing this? -->
+
+# Prior art
+[prior-art]: #prior-art
+
+Many other languages have package managers with support for private
+repositories. Not all will be listed here, but a select few as examples.
```

Comment:
I could use more info here:

- how do etch the other package managers do authentication?
- how are the specifics different?
- why did they do it differently?

Reply:
Of those examples, I know that Pip, Maven, and Composer use HTTP Basic, and NPM uses some not very well documented token mechanism.